pkg/cluster/rook_ceph.go

```diff
@@ -41,31 +41,44 @@ func init() {
 }
 
 // returns the number of nodes used for storage, which may be higher than the number of names passed in
-// if a node is currenlty not ready but has not been purged
+// if a node is currently not ready but has not been purged
 func (c *Controller) UseNodesForStorage(rookVersion semver.Version, names []string) (int, error) {
 	cluster, err := c.GetCephCluster(context.TODO())
 	if err != nil {
 		return 0, errors.Wrapf(err, "get CephCluster config")
 	}
+	var next []cephv1.Node
 	storageNodes := make(map[string]bool, len(cluster.Spec.Storage.Nodes))
 	for _, storageNode := range cluster.Spec.Storage.Nodes {
+		next = append(next, storageNode)
 		storageNodes[storageNode.Name] = true
 	}
 	changed := false
 	for _, name := range names {
 		if !storageNodes[name] {
 			c.Log.Infof("Adding node %q to CephCluster node storage list", name)
-			cluster.Spec.Storage.Nodes = append(cluster.Spec.Storage.Nodes, cephv1.Node{
+			next = append(next, cephv1.Node{
 				Name: name,
 			})
 			changed = true
 		}
 	}
 	if changed {
```

Comment:
Not a blocker either because that is out of the scope. But, just for share I think we need review and refac the controllers implementations in general. By looking a look on it shows that we are using controller runtime appropriated. 
- a) I could not find the watches features implemented accordingly for we are able to re-trigger the reconciliation when some event to create/update/delete an resource is trigged
- b) it seems that we are doing like a install something approach to run just once when we could ensure an idempotent solution, todo the operations into the reconcile and ensure that every time that something change that we watching change the reconciliation would be re-trigged) 
- c) We are not using status conditionals to keep track the status of the desired resources which can help us to troubleshooting see; 
- https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties
- https://maelvls.dev/kubernetes-conditions/

Reply:
Agreed but out of scope. Can you create a story in Replicated's internal Shortcut?

Reply:
I will. Thx.